source/extensions/filters/http/cache/cache_headers_utils.cc

```diff
@@ -0,0 +1,157 @@
+            << "only_if_cached: " << request_cache_control.only_if_cached << ", "
+            << "max_age: " << request_cache_control.max_age << ", "
+            << "min_fresh: " << request_cache_control.min_fresh << ", "
+            << "max_stale: " << request_cache_control.max_stale << "}";
+}
+
+// TODO: clean or remove
+std::ostream& operator<<(std::ostream& os, const ResponseCacheControl& response_cache_control) {
+  return os << "{"
+            << "must_validate: " << response_cache_control.must_validate << ", "
+            << "no_store: " << response_cache_control.no_store << ", "
+            << "no_transform: " << response_cache_control.no_transform << ", "
+            << "no_stale: " << response_cache_control.no_stale << ", "
+            << "public: " << response_cache_control._public << ", "
+            << "max_age: " << response_cache_control.max_age << "}";
+}
+
+SystemTime CacheHeadersUtils::httpTime(const Http::HeaderEntry* header_entry) {
+  if (!header_entry) {
+    return {};
+  }
+  absl::Time time;
+  const std::string input(header_entry->value().getStringView());
+
+  // Acceptable Date/Time Formats per
+  // https://tools.ietf.org/html/rfc7231#section-7.1.1.1
+  //
+  // Sun, 06 Nov 1994 08:49:37 GMT    ; IMF-fixdate
+  // Sunday, 06-Nov-94 08:49:37 GMT   ; obsolete RFC 850 format
+  // Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
+  static const auto& rfc7231_date_formats = *new std::array<std::string, 3>{
+      "%a, %d %b %Y %H:%M:%S GMT", "%A, %d-%b-%y %H:%M:%S GMT", "%a %b %e %H:%M:%S %Y"};
+  for (const std::string& format : rfc7231_date_formats) {
+    if (absl::ParseTime(format, input, &time, nullptr)) {
+      return ToChronoTime(time);
+    }
+  }
+  return {};
+}
+
+OptionalDuration CacheHeadersUtils::parseDuration(absl::string_view s) {
+  OptionalDuration duration;
+  // Strip quotation marks if any
+  if (!s.empty() && s.front() == '"' && s.back() == '"') {
+    s = s.substr(1, s.size() - 2);
+  }
+  long num;
+  if (absl::SimpleAtoi(s, &num) && num >= 0) {
+    // s is a valid string of digits representing a positive number
+    duration = std::chrono::seconds(num);
+  }
+  return duration;
+}
+
+std::tuple<absl::string_view, absl::string_view>
+CacheHeadersUtils::separateDirectiveAndArgument(absl::string_view full_directive) {
+  std::vector<absl::string_view> directive_parts =
+      absl::StrSplit(full_directive, absl::MaxSplits('=', 1));
+  absl::string_view directive = absl::StripAsciiWhitespace(directive_parts[0]);
+  absl::string_view argument =
+      directive_parts.size() > 1 ? absl::StripAsciiWhitespace(directive_parts[1]) : "";
+  return std::make_tuple(directive, argument);
+}
+
+RequestCacheControl CacheHeadersUtils::requestCacheControl(absl::string_view cache_control_header) {
+  RequestCacheControl request_cache_control;
+  std::vector<absl::string_view> directives = absl::StrSplit(cache_control_header, ',');
```

Comment:
How well does this code handle request directives with ',' in the argument? If the worst that can happen is an incorrect result, that should be ok, but please add a comment warning future maintainers to be careful.

Reply:
The worst is that this directive could be interpreted in a wrong way.
For example: `cache-control: private, max-age=5, max-stale=10,0, no-store`.
This will split to:
`private` , `max-age=5` , `max-stale=10`, `0` , `no-store`
The `0` will be ignored, and `max-stale` will be misinterpreted as 10.

Reply:
Which kinda make sense. 
A header like this is probably more likely to be intended as `max-stale=10` and another directive that was written incorrectly as `0`, rather than `max-stale=100` with a random injected comma.

Reply:
Added comment with an example.

Reply:
@capoferro it's not an intended behavior, but rather an output of a wrong input, which shouldn't matter. I mean if the behavior changes to max-stale not being parsed at all for example, that should not be a problem (as the input is ill-formed anyway). What do you think?

Reply:
Ensuring that we handle invalid input is also worthy of a test. Tests serve as living documentation of the intended behavior of a function.

Reply:
I already have some test cases of invalid inputs, will add one for this case.